galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker.go

```diff
@@ -0,0 +1,224 @@
+		return nil
+	}
+	// Discover the targeted workload and take note. Should normalize.
+	for _, target := range p.Targets {
+		fqdn := util.ConvertHostToFQDN(namespace, target.Name)
+
+		if len(target.Ports) == 0 {
+			// Policy targets all ports on workload
+			pc.workloadToMTLSMode[NewWorkload(fqdn)] = policyMode
+		}
+
+		for _, port := range target.Ports {
+			if port.GetName() != "" {
+				pc.workloadToMTLSMode[NewWorkloadWithPortName(fqdn, port.GetName())] = policyMode
+			} else if port.GetNumber() != 0 {
+				pc.workloadToMTLSMode[NewWorkloadWithPortNumber(fqdn, port.GetNumber())] = policyMode
+			} else {
+				// Unhandled case!
+				return fmt.Errorf("policy has a port with no name/number for target %s", target.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsServiceMTLSEnforced returns true if a workload requires incoming
+// connections to use MTLS, or false if MTLS is not a hard-requirement (e.g.
+// mode is permissive, peerIsOptional is true, etc). Only call this after adding
+// all policy resources in effect via AddPolicy or AddMeshPolicy.
+func (pc *PolicyChecker) IsServiceMTLSEnforced(w Workload) (bool, error) {
+	// TODO support understanding port name -> port number mappings
+	switch pc.workloadToMTLSMode[w] {
+	case strictMTLSExplicitlyEnabled:
+		return true, nil
+	case strictMTLSExplicitlyDisabled:
+		return false, nil
+	case strictMTLSUnset:
+		// Fall through switch case
+	default:
+		return false, fmt.Errorf("unknown strictMTLSMode: %v", pc.workloadToMTLSMode[w])
+	}
+
+	// Try checking if its enforced on any ports
+	workloadNoPort := NewWorkload(w.FQDN())
+	switch pc.workloadToMTLSMode[workloadNoPort] {
+	case strictMTLSExplicitlyEnabled:
+		return true, nil
+	case strictMTLSExplicitlyDisabled:
+		return false, nil
+	case strictMTLSUnset:
+		// Fall through switch case
+	default:
+		return false, fmt.Errorf("unknown strictMTLSMode: %v", pc.workloadToMTLSMode[workloadNoPort])
+	}
+
+	// Check if enforced on namespace
+	namespace, _ := util.GetResourceNameFromHost("", w.FQDN()).InterpretAsNamespaceAndName()
+	if namespace == "" {
+		return false, fmt.Errorf("unable to extract namespace from fqdn: %s", w.FQDN())
+	}
+	switch pc.namespaceToMTLSMode[namespace] {
+	case strictMTLSExplicitlyEnabled:
+		return true, nil
+	case strictMTLSExplicitlyDisabled:
+		return false, nil
+	case strictMTLSUnset:
+		// Fall through switch case
+	default:
+		return false, fmt.Errorf("unknown strictMTLSMode: %v", pc.namespaceToMTLSMode[namespace])
+	}
+
+	// Finally, defer to mesh level policy
+	return pc.meshHasStrictMTLSPolicy, nil
+}
+
+// doesPolicyEnforceMTLS is a helper function to determine whether or not a
+// Policy implies Strict MTLS mode.
+func doesPolicyEnforceMTLS(p *v1alpha1.Policy) bool {
+	if p.PeerIsOptional {
+		// Connection can still occur.
+		return false
+	}
+	hasStrictMTLSPolicy := false
+	for _, peer := range p.Peers {
+		mtlsParams, ok := peer.Params.(*v1alpha1.PeerAuthenticationMethod_Mtls)
```

Comment:
What happens if I have multiple `mTLS` blocks in peers for the same Policy object and one says enabled and other says disabled?

Reply:
Good question - with this implementation, only one will be considered authoritative.

A future analyzer here could complain specifically about this, but I think it'd be awkward to do the checking here (unless you'd rather just log something).

Reply:
with this implementation, only one will be considered authoritative. -> is that what Pilot does? We need to be consistent with Pilot here so let's test this case and mimic whatever pilot is doing.

Reply:
Agh, I misinterpretted your first comment. I thought you were talking about the case where multiple Policy objects targetted the same workload, but I see you're talking about a single Policy object with multiple MTLS configurations.

According to the documentation:
```
	// List of authentication methods that can be used for peer authentication.
	// They will be evaluated in order; the first validate one will be used to
	// set peer identity (source.user) and other peer attributes. If none of
	// these methods pass, request will be rejected with authentication failed error (401).
	// Leave the list empty if peer authentication is not required
```

If pilot does behave as described above, I think the policy checker should treat the service as 'StrictMTLSRequired' if _any_ mtls configuration requires strict. 

I'll update the implementation to do this.

Reply:
After diving into the comments, it looks like only the first mtls policy specified holds.

@jasonwzm is there any validation on Authentication policies that specifies that mtls isn't specified multiple times? How would we add something like that?